### PR TITLE
Addressed Fan models need to protect against negative fan power due to bad EMS override values

### LINKED ISSUE #7365: Fan models need to protect against negative fan power due to bad EMS override values
Issue overview
--------------
User file failed with a fatal psychrometric error due to very cold temperatures in the HVAC system. The problem comes from negative values set by EMS for fan pressure , which caused negative values for fan power and fan heat gain, leading to very cold node temperatures, which then cascaded until the psych error tripped.

All of the fan models need a trap for this. I'd be find with simply putting a max(power,0.0) around the first power values calculated in each fan model - just clip it silently without warning.

### Details
Some additional details for this issue (if relevant):
 - Version of EnergyPlus 9.1.0-08d2e308bb
 - Helpdesk ticket number 15206

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Defect file added in EnergyPlusDevSupport\DefectFiles
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by Nigusse**
Pull request overview
---------------------
Fixes #7365 

### Work Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
   - Defect: This pull request repairs a github defect issue.  The github issue should be referenced in the PR description
   - Performance: This pull request includes code changes that are directed at improving the runtime performance of EnergyPlus

### Review Checklist
This will not be exhaustively relevant to every PR.
 - [ ] Functional code review (it has to work!)
 - [ ] If defect, results of running current develop vs this branch should exhibit the fix
 - [ ] CI status: all green or justified
 - [ ] Performance: CI Linux results include performance check
 - [ ] Unit Test(s)
 - C++ checks:
   - [ ] Argument types
   - [ ] If any virtual classes, ensure virtual destructor included, other things
 - IDD changes:
   - [ ] Verify naming conventions and styles, memos and notes and defaults
   - [ ] Open windows IDF Editor with modified IDD to check for errors
   - [ ] If transition, add to input rules file for interfaces
   - [ ] If transition, add transition source
   - [ ] If transition, update idfs

### CODE REVIEW

File: tst/EnergyPlus/unit/Fans.unit.cc [hunk 1]
```diff
@@ -98,3 +98,156 @@ TEST_F(EnergyPlusFixture, Fans_FanSizing)
...
+    EXPECT_DOUBLE_EQ(Result_FanPower, thisFan.FanPower); // expects 300 W
+
+    // negative fan pressure rise set using EMS
+    thisFan.EMSFanPressureOverrideOn = true;
+    thisFan.EMSFanPressureValue = -300.0;
+    // simulate the fan with negative pressure rise
+    // set using fans EMS actuator for Pressure Rise
+    Fans::SimOnOffFan(FanNum);
+    Real64 Result2_FanPower = max(0.0, thisFan.MaxAirMassFlowRate * thisFan.EMSFanPressureValue / (thisFan.FanEff * thisFan.RhoAirStdInit));
+    EXPECT_DOUBLE_EQ(Result2_FanPower, thisFan.FanPower); // expects zero
```
Comment (Myoldmopar) [inline tst/EnergyPlus/unit/Fans.unit.cc:194]: :+1: makes sense based on the desired changes here.

File: src/EnergyPlus/HVACFan.cc [hunk 2]
```diff
@@ -866,8 +868,8 @@ namespace HVACFan {
                             localFanTotEff = m_fanTotalEff;
                             locHiSpeedFanRunTimeFrac = locFanRunTimeFraction;
                             m_fanRunTimeFractionAtSpeed[0] += locHiSpeedFanRunTimeFrac;
-                            m_fanPower +=
-                                locHiSpeedFanRunTimeFrac * m_maxAirMassFlowRate * localPressureRise[mode] / (localFanTotEff * m_rhoAirStdInit);
+                            m_fanPower += max(
```
Comment (Myoldmopar) [inline src/EnergyPlus/HVACFan.cc:871]: I understand the intent here, but seeing all these spread out has me concerned about a performance impact.  Is there a more central location we can check for this and have less max calls later on?  Probably not but I feel I should ask.

Comment (Nigusse) [reply src/EnergyPlus/HVACFan.cc:871]: One central location for Fan:SystemModel check is [here](https://github.com/NREL/EnergyPlus/blob/167109939_Issue7365/src/EnergyPlus/HVACFan.cc#L976). But, will not work for "SpeedControlMethod::Continuous" due to the code [here](https://github.com/NREL/EnergyPlus/blob/167109939_Issue7365/src/EnergyPlus/HVACFan.cc#L939-L943) used to update fan outlet air condition.

The other possibility is to reset EMS fan pressure rise value [here](https://github.com/NREL/EnergyPlus/blob/167109939_Issue7365/src/EnergyPlus/HVACFan.cc#L751) but I do not know how significant would the performance impact be until we test it.

Comment (Nigusse) [reply src/EnergyPlus/HVACFan.cc:871]: @Myoldmopar @mjwitte Thinking about it, adding a check for the EMS pressure rise only is executed only when the EMS variable EMSFanPressureOverrideOn is true. This reduces the computation effort, but negative fan power will not be captured if it happen for other reasons, for instance when the Fan(FanNum).EMSAirMassFlowValue is negative.

**Comment by Myoldmopar** [comment]
This still looks like a good set of changes.  I'm going to pull the code down, merge up with develop and re-run unit tests just to be sure it's all set.

**Comment by Myoldmopar** [comment]
Yep, everything looks good locally.  Merging.

**Pull request merged**